### PR TITLE
fix: support different tag association queries for COLUMN object types

### LINKED ISSUE #1201: snowflake_tag_association does not support column-level tagging
**Provider Version**

0.43.0

**Terraform Version**

1.2.1

**Describe the bug**

It is not possible to create a column tag association, even though it is a supported `object_type`. This should execute a `ALTER TABLE ... ALTER COLUMN` statement, but currently executes a `ALTER COLUMN`-statement, which is not valid.

**Expected behavior**

The resource should support passing database, schema, table (only supported for column) and name as separate attributes, or parse the attributes similarly to what is done for the `tag_id` attribute. When the `object_type` is "COLUMN", an `ALTER TABLE ... ALTER COLUMN` statement should be executed.

**Code samples and commands**

```
# COLUMN
resource "snowflake_tag_association" "tag_associations" {
    object_name     = "DB.SCHEMA.TABLE_NAME.TEST_COLUMN"
    object_type     = "COLUMN"
    tag_id          = "TAG_DB.TAG_SCHEMA.TAG_NAME"
    tag_value       = "TEST"
}
```

**Additional context**

The statement above executes `ALTER COLUMN "DB.SCHEMA.TABLE_NAME.TEST_COLUMN" SET TAG "TAG_DB"."TAG_SCHEMA"."TAG_NAME" = "TEST"` which gives a syntax error as `ALTER COLUMN ...` is not supported on Snowflake.

### PR DESCRIPTION
**Pull request by rmorris1218**
This is intended to close #1201 by implementing some logic to handle the slightly different query needed for `COLUMN` tag associations vs. other object types.

## Test Plan

* [ ] acceptance tests

* [ ] …

## References

*

### CODE REVIEW

**Comment by sfc-gh-swinkler** [comment]
```
    tag_association_test.go:45: 
        	Error Trace:	/home/runner/work/terraform-provider-snowflake/terraform-provider-snowflake/pkg/snowflake/tag_association_test.go:45
        	Error:      	Not equal: 
        	            	expected: "c"
        	            	actual  : "\"a\".\"b\".\"c\""
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-c
        	            	+"a"."b"."c"
        	Test:       	TestTableColumnName
```
Hello @rmorris1218! There is a minor issue with the test case TestTableColumnName. If you could please correct then we can get this PR merged.

File: pkg/snowflake/tag_association.go [hunk 1]
```diff
@@ -76,12 +88,22 @@ func NewTagAssociationBuilder(tagID string) *TagAssociationBuilder {
 
 // Create returns the SQL query that will set the tag on an object.
 func (tb *TagAssociationBuilder) Create() string {
-	return fmt.Sprintf(`ALTER %v %v SET TAG "%v"."%v"."%v" = '%v'`, tb.objectType, tb.objectIdentifier, tb.databaseName, tb.schemaName, tb.tagName, EscapeString(tb.tagValue))
+	if strings.ToUpper(tb.objectType) == "COLUMN" {
+		tableName, columnName := tb.GetTableAndColumnName()
+		return fmt.Sprintf(`ALTER TABLE %v ALTER COLUMN %v SET TAG "%v"."%v"."%v" = '%v'`, tableName, columnName, tb.databaseName, tb.schemaName, tb.tagName, EscapeString(tb.tagValue))
```
Comment (AndresPeralta) [inline pkg/snowflake/tag_association.go:93]: We should change the `ALTER COLUMN` with `MODIFY COLUMN`
```suggestion
		return fmt.Sprintf(`ALTER TABLE %v MODIFY COLUMN %v SET TAG "%v"."%v"."%v" = '%v'`, tableName, columnName, tb.databaseName, tb.schemaName, tb.tagName, EscapeString(tb.tagValue))
```

File: pkg/snowflake/tag_association.go [hunk 2]
```diff
@@ -76,12 +88,22 @@ func NewTagAssociationBuilder(tagID string) *TagAssociationBuilder {
...
+		return fmt.Sprintf(`ALTER %v %v SET TAG "%v"."%v"."%v" = '%v'`, tb.objectType, tb.objectIdentifier, tb.databaseName, tb.schemaName, tb.tagName, EscapeString(tb.tagValue))
+	}
 }
 
 // Drop returns the SQL query that will remove a tag from an object.
 func (tb *TagAssociationBuilder) Drop() string {
-	return fmt.Sprintf(`ALTER %v %v UNSET TAG "%v"."%v"."%v"`, tb.objectType, tb.objectIdentifier, tb.databaseName, tb.schemaName, tb.tagName)
+	if strings.ToUpper(tb.objectType) == "COLUMN" {
+		tableName, columnName := tb.GetTableAndColumnName()
+		return fmt.Sprintf(`ALTER TABLE %v ALTER COLUMN %v UNSET TAG "%v"."%v"."%v"`, tableName, columnName, tb.databaseName, tb.schemaName, tb.tagName)
```
Comment (AndresPeralta) [inline pkg/snowflake/tag_association.go:103]:
```suggestion
		return fmt.Sprintf(`ALTER TABLE %v MODIFY COLUMN %v UNSET TAG "%v"."%v"."%v"`, tableName, columnName, tb.databaseName, tb.schemaName, tb.tagName)
```

File: pkg/snowflake/tag_association_test.go [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+func TestTagAssociation(t *testing.T) {
+	tests := []TagAssociationTest{
+		{
+			Builder:        NewTagAssociationBuilder("test_db|test_schema|sensitive").WithObjectIdentifier(`"test_db"."test_schema"."test_table"`).WithObjectType("TABLE").WithTagValue("true"),
+			ExpectedCreate: `ALTER TABLE "test_db"."test_schema"."test_table" SET TAG "test_db"."test_schema"."sensitive" = 'true'`,
+			ExpectedDrop:   `ALTER TABLE "test_db"."test_schema"."test_table" UNSET TAG "test_db"."test_schema"."sensitive"`,
+		},
+		{
+			Builder:        NewTagAssociationBuilder("test_db|test_schema|sensitive").WithObjectIdentifier(`"test_db"."test_schema"."test_table.important"`).WithObjectType("COLUMN").WithTagValue("true"),
+			ExpectedCreate: `ALTER TABLE "test_db"."test_schema"."test_table" ALTER COLUMN important SET TAG "test_db"."test_schema"."sensitive" = 'true'`,
```
Comment (AndresPeralta) [inline pkg/snowflake/tag_association_test.go:24]: same here:
```suggestion
			ExpectedCreate: `ALTER TABLE "test_db"."test_schema"."test_table" MODIFY COLUMN important SET TAG "test_db"."test_schema"."sensitive" = 'true'`,
```

File: pkg/snowflake/tag_association_test.go [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+			ExpectedDrop:   `ALTER TABLE "test_db"."test_schema"."test_table" ALTER COLUMN important UNSET TAG "test_db"."test_schema"."sensitive"`},
```
Comment (AndresPeralta) [inline pkg/snowflake/tag_association_test.go:25]:
```suggestion
			ExpectedDrop:   `ALTER TABLE "test_db"."test_schema"."test_table" MODIFY COLUMN important UNSET TAG "test_db"."test_schema"."sensitive"`},
```

File: pkg/snowflake/tag_association_test.go [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+		{
+			Builder:        NewTagAssociationBuilder("OPERATION_DB|SECURITY|PII_2").WithObjectIdentifier(`"OPERATION_DB"."SECURITY"."test_table.important"`).WithObjectType("COLUMN").WithTagValue("true"),
+			ExpectedCreate: `ALTER TABLE "OPERATION_DB"."SECURITY"."test_table" ALTER COLUMN important SET TAG "OPERATION_DB"."SECURITY"."PII_2" = 'true'`,
```
Comment (AndresPeralta) [inline pkg/snowflake/tag_association_test.go:28]:
```suggestion
			ExpectedCreate: `ALTER TABLE "OPERATION_DB"."SECURITY"."test_table" MODIFY COLUMN important SET TAG "OPERATION_DB"."SECURITY"."PII_2" = 'true'`,
```

File: pkg/snowflake/tag_association_test.go [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+			ExpectedDrop:   `ALTER TABLE "OPERATION_DB"."SECURITY"."test_table" ALTER COLUMN important UNSET TAG "OPERATION_DB"."SECURITY"."PII_2"`},
```
Comment (AndresPeralta) [inline pkg/snowflake/tag_association_test.go:29]:
```suggestion
			ExpectedDrop:   `ALTER TABLE "OPERATION_DB"."SECURITY"."test_table" MODIFY COLUMN important UNSET TAG "OPERATION_DB"."SECURITY"."PII_2"`},
```

**Review by AndresPeralta — CHANGES_REQUESTED** [review]
We are calling ALTER TABLE x ALTER COLUMN y when should be:
ALTER TABLE x MODIFY COLUMN y

**Comment by rmorris1218** [comment]
> We are calling ALTER TABLE x ALTER COLUMN y when should be: ALTER TABLE x MODIFY COLUMN y

Thanks for the heads up! Updated the syntax for these.

**Comment by ldardick** [comment]
@sfc-gh-swinkler  any chance this can be reviewed by the Snowflake team? It's a blocker for us and the requested changes seems to be already applied by the owner. Let me know if I can help  with something

**Review by AndresPeralta — APPROVED** [review]

**Comment by sfc-gh-swinkler** [comment]
Sorry this slipped through the cracks. This looks okay to me. what i will do is merge it and resolve any workflow errors myself and get in the next release.

**Pull request merged**